Add EIP: RowDAS - Distributed Blob Reconstruction (#12118)
ethereum/EIPs: EIPS/eip-8371.md

---

> eip: 9999

Reviewer: ```suggestion
eip: 8371
```

Please also edit the filename and the `discussions-to`, thanks 😄 👍

title: RowDAS - Distributed Blob Reconstruction
description: Distribute reconstruction load in the network through row-level cell messaging.
author: Csaba Kiraly (@cskiraly)

> discussions-to: https://ethereum-magicians.org/t/eip-rowdas-distributed-blobspace-reconstruction/29320

Reviewer 2: ```suggestion
discussions-to: https://ethereum-magicians.org/t/eip-8371-rowdas-distributed-blobspace-reconstruction/29320
```
Updated Eth Magicians title with assigned number

status: Draft
type: Standards Track
category: Networking
created: 2026-08-05
requires: 7594, 8136
---

## Abstract

> PeerDAS requires supernodes to provide reconstruction, and this puts a high burden on supernodes that scales linearly with blob count. RowDAS enables distributed blobspace reconstruction using partial-message-based row topics, allowing all nodes to contribute to reconstruction, while significantly reducing the load on supernodes, leading to a more efficient and more resilient DAS construct.

Reviewer: ```suggestion
PeerDAS (Peer Data Availability Samping, [EIP-7594](./eip-7594.md) requires supernodes to provide reconstruction, and this puts a high burden on supernodes that scales linearly with blob count. RowDAS enables distributed blobspace reconstruction using partial-message-based row topics, allowing all nodes to contribute to reconstruction, while significantly reducing the load on supernodes, leading to a more efficient and more resilient DAS construct.
```

Introduces the reader to the DAS acronym and directly links to the intended (?) EIP. 

The acronym is rather strict, the link to PeerDAS EIP is not something I'm married to since you already link to it in motivation

Reviewer: Linking to the EIP first and then in the sentence later explaining the DAS acronym is also fine. Ensure the acronym is defined though, this saves the reader a hop or some time figuring out what DAS is if they do not know or do not remember.

## Motivation

[EIP-7594](./eip-7594.md) PeerDAS was designed with a simple but powerful-enough erasure coding based reconstruction model where any node receiving at least half of the 128 columns should reconstruct the whole extended blob content belonging to a block. As the number of blobs grows, however, the reconstruction burden on every supernode also grows linearly with blob count.

Moreover, supernodes execute largely redundant work: each one of them reconstructing all missing blobs, without the means to distribute this work efficiently in the network.

This EIP introduces distributed blobspace reconstruction, where different nodes prioritize the reconstruction of different parts of the blobspace, leading to a faster, less CPU-intensive, and more resilient construct.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).

> The EIP introduces new GossipSub topics, changes to the rules of reconstruction, and a few minor changes to how current column topics operate.

Reviewer: The term "supernode" is used a lot here. This might be a basic networking term. I nevertheless think it is good to define what a supernode is. Because I could imagine it either being a node with a lot of connections, or a powerful node (which not necessarily has a lot of connections but does specific tasks). Or maybe it is another definition.

### Column topics

Regarding column topics [EIP-7594](./eip-7594.md) already mandates the following:

> Once the node obtains a column through reconstruction, the node MUST expose the new column as if it had received it over the network. If the node is subscribed to the subnet corresponding to the column, it MUST send the reconstructed DataColumnSidecar to its topic mesh neighbors. If instead the node is not subscribed to the corresponding subnet, it SHOULD still expose the availability of the DataColumnSidecar as part of the gossip emission process.

This is extended to allow cell-level operation towards peers that support it using the following rules:

> - Prior to reconstructing, a node MAY also use advertisements received as part of the Gossipsub "fanout" mechanism to collect relevant cells from other peers across column subnets it is not subscribed to.

Reviewer: ```suggestion
- Prior to reconstructing, a node MAY also use advertisements received as part of the GossipSub "fanout" mechanism to collect relevant cells from other peers across column subnets it is not subscribed to.
```

I do not know what the correct styling is, but use the same style (GossipSub / Gossipsub / gossipsub) through the doc (these three variants are used)

- After reconstruction, a node SHOULD use the Gossipsub "fanout" mechanism to provide cells from the reconstructed blob to peers across column subnets it is not subscribed to. A node MAY choose to only advertise to a random subset of these columns rather than all columns. This allows the node to provide another path for cell dissemination to the network. A node MAY choose to delay these fanout messages in order to conserve bandwidth by not competing with other nodes who are subscribed to the column topic, and can provide the cell instead.

### Row topics

Similar to column subnets, we introduce new row subnets: `data_row_{subnet_id}`. These resemble, but are not to be confused with, the deprecated `blob_sidecar_{subnet_id}` topics. Their properties:

> - A row subnet MUST use Cell-Level Deltas without eager push. Like for Cell-Level Deltas in column subnets, the GroupID for a message in the row subnet is the block root. Since cells might arrive from three different sources (getBlobs, columns, rows) a node MAY choose to delay the request of cells from rows.

Reviewer: To make the Specification section more self-contained and readable in one-go, link to what you mean with Cell-Level Deltas, by linking to the EIP ([EIP-8136](./eip-8136.md) ?)

```suggestion
- A row subnet MUST use Cell-Level Deltas ([EIP-8136](./eip-8136.md)) without eager push. Like for Cell-Level Deltas in column subnets, the GroupID for a message in the row subnet is the block root. Since cells might arrive from three different sources (getBlobs, columns, rows) a node MAY choose to delay the request of cells from rows.
```

> - We make the number of subnet_ids equal to maximum blob count. However, to distribute load evenly in the network, we introduce a mapping function that rotates which blob goes to which subnet_id in which slot (exact permutation function TBD).

Reviewer: Style: (remove "we")

```suggestion
- The number of subnet_ids equals the maximum blob count. However, to distribute load evenly in the network, we introduce a mapping function that rotates which blob goes to which subnet_id in which slot (exact permutation function TBD).
```

(to move to Review TBD has to be defined)

> - Nodes subscribe to a single row subnet, derived with a pseudo-random function from their node ID similar to how it is done for custody columns; however, there is no custody requirement.

Reviewer: This sentence reads a bit vague. SHOULD nodes or MUST nodes subscribe to this subnet? "Similar to how it is done for custody columns" - how? what? If this is not known yet, add it as TBD.

(I am a novice in networking so this might be super trivial, but try to specify it also for novices like me 😄 )

A peer MAY limit the number of cells it serves a peer on the row subnet to just half of all the cells, as the rest can be reconstructed.

As a node receives cells from any source (either from row-subnets, column-subnets, or getBlobs), it SHOULD send updated bitmap states to its peers. A node MAY choose to debounce these updates.

### Reconstruction

A node, even if not a supernode, SHOULD collect at least 64 cells on its designated row and expose these in updated bitmap states to its peers. It MAY also decide to reconstruct the whole row. If it does, it SHOULD send its updated bitmap state to its peers.

> Similar to PeerDAS, a supernode (in this context any node with 64 or more column subscriptions) MUST contribute to reconstruction. However, reconstruction becomes a two-phase process:

Reviewer: This redefines the term supernode, also with the normative MUST statement, so this is confusing as the reader might now have to re-apply these rules to what previously was assigned in their perspective as "supernode". Please define supernode at the start of this section so it is clear what is meant with this throughout this spec.

- 1st phase: a supernode MUST reconstruct its designated row subnet, and it MUST send updated bitmap states to its peers. A small random delay (recommended range TBD) before reconstruction is allowed to desynchronise nodes in the network and reduce overall load.

> - 2nd phase: after a slightly longer random delay (recommended range TBD), during which cells are collected from getBlobs, columns, and rows, a supernode MUST do a second reconstruction phase, reconstructing all missing rows, and sharing the results as defined above.

Reviewer: ```suggestion
- 2nd phase: after a slightly longer random delay (recommended range TBD), during which cells are collected from `getBlobs`, columns, and rows, a supernode MUST do a second reconstruction phase, reconstructing all missing rows, and sharing the results as defined above.
```

This EIP corrects some of these shortcomings, making sure supernodes are not doing (as much) useless work, and reconstruction is possible (although not yet mandated) even without supernodes.

### What it is not

This EIP is not FullDAS. It does not introduce sub-linear (cell-level) sampling. The bandwidth requirement of sampling nodes is still proportional to the number of blobs.

It also does not introduce column-wise encoding, so protection and reconstruction is still only along the row axis.

Finally, it does not directly help L2 nodes retrieve individual blobs (although there are possible extensions in that direction). However, it helps them run supernodes with fewer resources, leading to a net gain.

### Possible extension to retrieve individual blobs

With the introduction of PeerDAS, L2 nodes have the problem that retrieving a specific blob from a CL client requires them to either be a supernode, or to download it on request through columns. By introducing row topics with allocation rooted in the nodeID, it is easier for nodes to identify which node they can download the relevant blob from. We reserve this for further consideration.

### Design decisions

> - why are non-supernodes part of the row-topics?

Reviewer: Formatting nit: these design-decision bullets read as lowercase question fragments. Suggest `###` sub-headings or bold sentence-case questions ("**Why are non-supernodes part of the row topics?**") for readability on the rendered page.

This is to enable the possibility of reconstructing without supernodes. The additional traffic of one row, most probably already suppressed by getBlobs, is worth it in our opinion.

- why are non-supernodes not required to reconstruct?

While mandated reconstruction would be desirable from the perspective of not relying on supernodes at all, this would introduce extra CPU load on nodes, and we try to avoid this in this phase. It is still recommended to activate this reconstruction path with a delay as a resilience mechanism.

- why only a single row, and why is it not dependent on custody?

> The current mainnet has approx. 12K nodes of which 3K are supernodes. The latter is much more than what we expected initially. With current and planned blob counts, even a single row creates abundant overlap.

Reviewer: Nit: date the measurements ("approx. 12K nodes of which 3K are supernodes") — e.g. "as of August 2026" — so the figure stays interpretable later.

## Backwards Compatibility

Row topics are limited to peers that have libp2p gossipsub implementations supporting Cell-level Deltas. The portion of peers that supports the extension is already reaching considerable numbers on mainnet, even before Glamsterdam. We expect the majority of peers to support it after the Glamsterdam fork and [EIP-8136](./eip-8136.md). For peers that do not support the extension, getBlobs and column topics are still fully available.

## Security Considerations

The EIP changes DAS networking, but it does not change the custody allocation and the probabilistic guarantees of PeerDAS.

New gossipsub topics might introduce new attack vectors. Since row distribution is a new additional recovery path, and the old paths are mainly intact, it is not expected that this adversely affects the system, except for a relatively small traffic overhead of bitmap-based signaling.
An exception to this is the two-phase reconstruction process. Here the 2nd phase, the full reconstruction, is explicitly delayed. This delay is, however, something implementations already practice, and it is a one-time (instead of hop-by-hop) delay.

> Since row topic count is bound to the maximum blob count, changes at BPO fork boundaries should be handled. Details to be defined.

Reviewer: Please expand BPO (Block Parameter Only) and (maybe?) link to EIP for clarity

```suggestion
Since row topic count is bound to the maximum blob count, changes at Block Parameter Only fork ([EIP-7982](./eip-7982.md)) boundaries should be handled. Details to be defined.
```
